### PR TITLE
Update Trilinos: Add Mesquite as optional package

### PR DESCRIPTION
**Pull request by jppelteret**
Also fix a build error that occurs when using an old compiler.

### CODE REVIEW

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 1]
```diff
@@ -342,6 +361,7 @@ class Trilinos(CMakePackage):
     patch('xlf_tpetra.patch', when='@12.12.1%xl')
     patch('xlf_tpetra.patch', when='@12.12.1%xl_r')
     patch('xlf_tpetra.patch', when='@12.12.1%clang')
+    patch('fix_clang_errors_12_18_1.patch', when='@12.18.1%clang')
```
Comment (aprokop) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Have you explored which versions should this patch be applied to?

**Comment by scheibelp** [comment]
Is there an upstream patch corresponding to `fix_clang_errors_12_18_1.patch` (e.g. something submitted to https://github.com/trilinos/Trilinos)?

**Comment by jppelteret** [comment]
> Is there an upstream patch corresponding to fix_clang_errors_12_18_1.patch (e.g. something submitted to https://github.com/trilinos/Trilinos)?

No, I've not submitted anything upstream. I'd be happy for it to find its way there, but I'm not going to submit the patch myself. Sorry, I know its annoying to hear that but severe time constraints limit which side tasks (no matter how small) I'm willing to take on nowadays.

**Comment by jppelteret** [comment]
Thank you @aprokop and @adamjstewart for the review!

File: var/spack/repos/builtin/packages/trilinos/package.py (hunk 1 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/trilinos/package.py]: > Should I limit the scope of the patch a bit more (i.e. only for my ancient compiler, apple-clang 8.0.0

Yes please (something like `when='12.8.1%clang@:8'` should work). Could you also mention what version of Mac OS you are running on?

Thanks!

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/trilinos/package.py]: This patch seems pretty innocuous, and will likely be needed for other versions/compilers. I would go so far as to include it even for non-Clang compilers, but that's just me.

Comment (jppelteret) [reply var/spack/repos/builtin/packages/trilinos/package.py]: @scheibelp I'm running 10.11 (El Capitan). Super old... Wouldn't `when='12.8.1%clang@:8'` also target clang built through spack, for example? Is there any way to restrict it to just apple clang, like `when='12.8.1%clang@:8 platform=darwin'` perhaps?

> This patch seems pretty innocuous

This is true. I think that it just needs a C++11 compliant compiler, but nowadays Trilinos requires C++11 to be built AFAIK. This is what I think I would have liked to have done, but I didn't want to appear overly zealous about patching.

Comment (jppelteret) [reply var/spack/repos/builtin/packages/trilinos/package.py]: I see that there's a set of comments in the [cuda package](https://github.com/spack/spack/blob/4af630308619afb40b1de89d1f7bac0f57b70a38/lib/spack/spack/build_systems/cuda.py#L124-L132) that seem to reference the same idea of differentiating between Apple XCode clang and LLVM clang (more specifically, the difficulty in distinguishing which is the clang that's building this package on MacOS). What is the preferred way forward?

**Comment by scheibelp** [comment]
I have submitted https://github.com/trilinos/Trilinos/issues/6761, which refers to this PR. Looks good to merge - thanks!

**Pull request merged**